docs: Provision a cert with the ACME DNS-01 challenge via Certbot + Cloudflare (#2968)
docker-mailserver/docker-mailserver: docs/content/config/security/ssl.md

        --volume "${PWD}/docker-data/certbot/logs/:/var/log/letsencrypt/" \
        --volume "${PWD}/docker-data/certbot/secrets/:/tmp/secrets/certbot/"
        certbot/dns-cloudflare \
        certonly --dns-cloudflare --dns-cloudflare-credentials /tmp/secrets/certbot/cloudflare.ini -d mail.example.com
      ```

6. Run the service to provision a certificate:

    ```sh
    docker-compose run certbot-cloudflare
    ```

7. You should see the following log output:

    ```log
    Saving debug log to /var/log/letsencrypt/letsencrypt. log | Requesting a certificate for mail.example.com
    Waiting 10 seconds for DNS changes to propagate
    Successfully received certificate.
    Certificate is saved at: /etc/letsencrypt/live/mail.example.com/fullchain.pem
    Key is saved at: /etc/letsencrypt/live/mail.example.com/privkey.pem
    This certificate expires on YYYY-MM-DD.
    These files will be updated when the certificate renews.
    NEXT STEPS:
    - The certificate will need to be renewed before it expires. Certbot can automatically renew the certificate in background, but you may need to take steps to enable that functionality. See https://certbot.org/renewal structions.

>     ```
> 
> !!! tip "(Optional) Renewing certificate"

Reviewer: ```suggestion
    ```

After completing the steps above, your certificate should be ready to use.

??? tip "Renewing a certificate (Optional)"
```

>     Since the above example only implements how to provision a certificate, since generated certificate will expire in 90 days, you might want to implement renew feature here.
> 
>     In the following example, we will modify our created `docker-compose.yml` and create a container which will use for renewing the certificate
> 
>     First, we modify our `docker-compose.yml` to add a service called `certbot-cloudflare-renew`:

Reviewer: ```suggestion
    We've only demonstrated how to provision a certificate, but it will expire in 90 days and need to be renewed before then.

    In the following example, add a new service (`certbot-cloudflare-renew`) into `docker-compose.yml` that will handle certificate renewals:
```

    ```yml
    services:
      certbot-cloudflare-renew:
        image: certbot/dns-cloudflare:latest
        command: renew --dns-cloudflare --dns-cloudflare-credentials /run/secrets/cloudflare-api-token
        volumes:
          - ./docker-data/certbot/certs/:/etc/letsencrtypt/
          - ./docker-data/certbot/logs/:/var/log/letsencrypt/
        secrets:
          - cloudflare-api-token

>     If needed, run the service to renew a certificate:

Reviewer: ```suggestion
    ```

    You can manually run this service to renew the cert within 90 days:
```

    
    ```sh
    docker-compose run certbot-cloudflare-renew
    ```

    You should see the following output
    (The following log was generated with `--dry-run` options)

>     ```log
>      Saving debug log to /var/log/letsencrypt/letsencrypt.log
>     
>      - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
>      Processing /etc/letsencrypt/renewal/mail.example.com.conf
>      - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
>      Account registered.
>      Simulating renewal of an existing certificate for brtest.club
>      Waiting 10 seconds for DNS changes to propagate
>     
>      - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
>      Congratulations, all simulated renewals succeeded:
>        /etc/letsencrypt/live/mail.example.com/fullchain.pem (success)
>      - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
>     ``````

Reviewer: ```suggestion
    ```log
    Saving debug log to /var/log/letsencrypt/letsencrypt.log
    
    - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
    Processing /etc/letsencrypt/renewal/mail.example.com.conf
    - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
    Account registered.
    Simulating renewal of an existing certificate for mail.example.com
    Waiting 10 seconds for DNS changes to propagate
    
    - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
    Congratulations, all simulated renewals succeeded:
      /etc/letsencrypt/live/mail.example.com/fullchain.pem (success)
    - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
    ```
```

>     Then we add auto-renew certificate script into `crontab`
>     (Here we let certbot check if the certificate need renew every day)

Reviewer: ```suggestion
    It is recommended to automate this renewal via a task scheduler like a _systemd timer_ or in `crontab`
    (`crontab` example: Checks every day if the certificate should be renewed)
```

    ```sh
    0 0 * * * docker-compose -f PATH_TO_YOUR_DOCKER_COMPOSE_YML up certbot-cloudflare-renew
    ```

> 
> After completing the steps above, your certificate should be ready to use.

Reviewer: ```suggestion
```

    Certbot does support [alternative certificate providers via the `--server`][certbot::custom-ca] option. In most cases you'll want to use the default _Let's Encrypt_.

#### Example using `certbot-dns-cloudflare` with Docker { data-toc-label='certbot-dns-cloudflare with Docker' }

If for some reason, you can't expose your mail server's port 80 and port 443 (known as the `HTTP-01` challenge), you can consider obtaining a certificate by the `DNS-01` challenge.

First, you need to obtain a Cloudflare API token

1. Login into your Cloudflare dashboard.
2. Navigate to MyProfile -> API Tokens.
3. Create token a using the `Edit zone DNS` template since certbot [requires](https://certbot-dns-cloudflare.readthedocs.io/en/stable/#credentials) `ZONE:DNS:Edit` permission.

!!! Zone resource configuration

    Remember to choose a specific TLD, and create a token that allows access to all zone's resources is insecure here.

4. Create `cloudflare.ini` with the following content:

    ```dosini
    dns_cloudflare_api_token = YOUR_CLOUDFLARE_TOKEN_HERE
    ```

5. Save it to the `.secrets` folder or anywhere you like. (Make sure to restrict its permission with the `chmod 600` command

> 6. Add the following code into `docker-compose.yml`
> 
>     ```yaml
>     services:
>       mailserver:
>         environments:
>           # Set SSL certificate type.
>           - SSL_TYPE=letsencrypt
>         volumes:
>           # We need to mount the cert generated by certbot into mail server.
>           - ./docker-data/certbot/certs/:/etc/letsencrypt
>       certbot-cloudflare:
>         image: certbot/dns-cloudflare:latest
>         container_name: certbot-cloudflare
>         volumes:
>           - ./docker-data/certbot/certs/:/etc/letsencrypt/
>           - ./docker-data/certbot/logs/:/var/log/letsencrypt/
>           - ./.secrets:/.secrets
>         command: certonly --dns-cloudflare --dns-cloudflare-credentials /.secrets/certbot/cloudflare.ini -d mail.example.com
>     ```
> 
>     Alternative using the `docker run` command
> 
>       ```sh
>       docker run \
>         --volume "${PWD}/docker-data/certbot/certs/:/etc/letsencrypt/" \
>         --volume "${PWD}/docker-data/certbot/logs/:/var/log/letsencrypt/" \
>         --volume "${PWD}/.secrets:/.secrets"
>         certbot/dns-cloudflare \
>         certonly --dns-cloudflare --dns-cloudflare-credentials /.secrets/certbot/cloudflare.ini -d mail.example.com
>       ```

Reviewer: Apart from changing the list number to `5`, Revised a volume mount comment + `container_name` was removed.

Additionally changed the `docker-compose` config to use the [`secrets` feature](https://docs.docker.com/compose/compose-file/#secrets), although without Swarm only the file mount mode is available, and the long syntax features to adjust permissions and ownership don't work.. so I'm not sure how helpful that is as it's effectively a volume bind mount :sweat_smile: (_the `docker run` example also can't use it_).

So I'm open to dropping the `secrets` change, although I prefer it for being more explicit about content and separated from other volume mounts. The concerns about those gotcha's have been inlined into the docs example.

```suggestion
5. Your `docker-compose.yml` should include the following:

    ```yaml
    services:
      mailserver:
        environments:
          # Set SSL certificate type.
          - SSL_TYPE=letsencrypt
        volumes:
          # Mount the cert folder generated by Certbot into mail-server:
          - ./docker-data/certbot/certs/:/etc/letsencrypt/:ro

      certbot-cloudflare:
        image: certbot/dns-cloudflare:latest
        command: certonly --dns-cloudflare --dns-cloudflare-credentials /run/secrets/cloudflare-api-token -d mail.example.com
        volumes:
          - ./docker-data/certbot/certs/:/etc/letsencrypt/
          - ./docker-data/certbot/logs/:/var/log/letsencrypt/
        secrets:
          - cloudflare-api-token

    # Docs: https://docs.docker.com/engine/swarm/secrets/#use-secrets-in-compose
    # WARNING: In compose configs without swarm, the long syntax options have no effect,
    # Ensure that you properly `chmod 600` and `chown 0:0` the file on disk. Effectively treated as a bind mount.
    secrets:
      cloudflare-api-token:
        file: ./docker-data/certbot/secrets/cloudflare.ini
    ```

    Alternative using the `docker run` command (`secrets` feature is not available):

      ```sh
      docker run \
        --volume "${PWD}/docker-data/certbot/certs/:/etc/letsencrypt/" \
        --volume "${PWD}/docker-data/certbot/logs/:/var/log/letsencrypt/" \
        --volume "${PWD}/docker-data/certbot/secrets/:/tmp/secrets/certbot/"
        certbot/dns-cloudflare \
        certonly --dns-cloudflare --dns-cloudflare-credentials /tmp/secrets/certbot/cloudflare.ini -d mail.example.com
      ```
```

> 7. Turn it on to generate certificate.

Reviewer: ```suggestion
6. Run the service to provision a certificate:
```

    ```sh

>     docker-compose up certbot-cloudflare

Reviewer: Is this meant to run in the background to handle renewals at all? Or should the user manually invoke regularly?

If it's only intended for running manually `docker-compose run certbot-cloudflare` might be more suitable?

Author: Thanks for these suggestions, sorry I had a business trip yesterday.  
Generate certificate via certbot-cloudflare should be run manually since the part of renew certificate requires another action

Author: I've applied changes in [b32e713](https://github.com/docker-mailserver/docker-mailserver/pull/2968/commits/b32e71377769710093075156a04b2c0820b83b23), thanks for advice

    ```

> 8. You should see the following log output.

Reviewer: ```suggestion
7. You should see the following log output:
```

> 8. You should see the following log output.

Reviewer: ```suggestion
7. You should see the following log output:
```

    ```log
    Saving debug log to /var/log/letsencrypt/letsencrypt. log | Requesting a certificate for mail.example.com
    Waiting 10 seconds for DNS changes to propagate
    Successfully received certificate.
    Certificate is saved at: /etc/letsencrypt/live/mail.example.com/fullchain.pem
    Key is saved at: /etc/letsencrypt/live/mail.example.com/privkey.pem
    This certificate expires on YYYY-MM-DD.
    These files will be updated when the certificate renews.
    NEXT STEPS:
    - The certificate will need to be renewed before it expires. Certbot can automatically renew the certificate in background, but you may need to take steps to enable that functionality. See https://certbot.org/renewal structions.
    ```

> After complete the steps above, your certificate should be ready to use.

Reviewer: ```suggestion
After completing the steps above, your certificate should be ready to use.
```

        volumes:
          - ./docker-data/certbot/certs/:/etc/letsencrypt/
          - ./docker-data/certbot/logs/:/var/log/letsencrypt/
        secrets:
          - cloudflare-api-token

    # Docs: https://docs.docker.com/engine/swarm/secrets/#use-secrets-in-compose
    # WARNING: In compose configs without swarm, the long syntax options have no effect,
    # Ensure that you properly `chmod 600` and `chown 0:0` the file on disk. Effectively treated as a bind mount.
    secrets:
      cloudflare-api-token:
        file: ./docker-data/certbot/secrets/cloudflare.ini
    ```

    Alternative using the `docker run` command (`secrets` feature is not available):

      ```sh
      docker run \
        --volume "${PWD}/docker-data/certbot/certs/:/etc/letsencrypt/" \
        --volume "${PWD}/docker-data/certbot/logs/:/var/log/letsencrypt/" \
        --volume "${PWD}/docker-data/certbot/secrets/:/tmp/secrets/certbot/"
        certbot/dns-cloudflare \
        certonly --dns-cloudflare --dns-cloudflare-credentials /tmp/secrets/certbot/cloudflare.ini -d mail.example.com
      ```

6. Run the service to provision a certificate:

    ```sh
    docker-compose run certbot-cloudflare

>     ```

Reviewer: As you've mentioned this only allows to provision the certificate and different steps are required for renewal. Would you like to inform the user about this and how to go about renewal?

Presently pairing the service in a compose config next to the DMS service would run that service and command each time the mail-server is brought up with a `docker-compose up -d`, is that going to be a concern? Should it be treated as a separate compose file?

I've not reviewed the other guides on this page, they may have a similar concern but not one you need to worry about :+1:

---

I [came across a very similar step-by-step guide](https://www.nodinrogers.com/post/2022-03-10-certbot-cloudflare-docker/#3-create-the-docker-composeyml-file) that shows that renewal command is pretty much the same but includes `renew` instead of `certonly`, they also have a few extra options in their commands (_also via a docker-compose file_) that might be worthwhile.

It seems common to just setup a cron or systemd timer on the host to schedule renewals with Certbot?

---

I'm not blocking approval on this, but it would be appreciated if addressed :sweat_smile:

Author: Sure, it'd be better if we add that part about renewal, gonna working on this part later and expect to create a commit later today

Author: I've completed add documentation for renewing certificate in [8560bdf](https://github.com/docker-mailserver/docker-mailserver/pull/2968/commits/8560bdf7f65c177a1a207104c2841eb2cad6c2be), waiting for github-action to complete building preview
